contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+  where no types or functions have definitions. Example:
+
+```ocaml
+// An abstract contract
+contract VotingType = {
+  public stateful let vote : string => unit;
+};
+```
+
+- A contract instance is an entity living on the block chain (or in a state channel). Each
+instance is associated with a particular contract implementation (at
+least from the view of high-level language; the low-level details of
+how to store and check API information on the chain is still to be
+worked out).
+- A contract may define a type state encapsulating its local
+  state. The state must be initialised when instantiating a
+  contract. This is done by the init function which can take arbitrary
+  arguments and is called on contract instance creation.
+- Contracts have a public API, comprising the functions and types
+  annotated with the public keyword. These can be used from outside
+  the contract. Functions and types with no annotation (or internal)
+  are part of the internal API and can be used by contracts inheriting
+  (see below) the given contract. Functions and types annotated with
+  private can only be used locally.
+- Contracts can inherit one (or more?) other contract(s). In this case
+  the public functions (and types) of the inherited contract are
+  included in the public API and internal functions are included in
+  the internal API of the current contract. The state of the contract
+  contains the states of all inherited contracts as well as the local
+  state. However, the state of an inherited contract cannot be
+  accessed other than through internal functions defined by that
+  contract.
+- Open question: should we allow overriding defined functions? I would
+  suggest no, but there might be compelling use cases that I'm
+  missing.
+
+### Mutable state
+
+Ring does not have arbitrary mutable state, but only a limited form of
+state associated with each contract instance.
+
+- Each contract defines a state type encapsulating its mutable state.
+- The value of the state is accessible from inside the contract
+  through an implicitly bound variable state.
+- State updates are performed by calling a function put : state =>
+  unit (possibly with a better name).
+- Aside from the put function (and similar functions for transactions
+  and events), the contract language is purely functional.
+- Functions modifying the state need to be annotated with the stateful keyword.
+
+To make it convenient to update parts of a deeply nested state Ring
+provides special special syntax for map/record updates.  Open
+question: we likely want to make it possible to have immutable state
+(parameters). Keep separate from mutable state or annotate certain
+fields as immutable?
+
+### Types
+Ring has the following types:
+
+| Type    | Description                     | Example
+| ------- | ------------------------------- | -------:
+| uint    | A 256 bit integer               | ```42```
+| int     | A 256 bit 2-complement integer  | ```-1```
```

Comment:
Is there a good reason to have 256-bit integers rather than arbitrary precision integers?

Reply:
Do we absolutely need both signed and unsigned integers? Overloading arithmetic operations will complicate type checking and compilation.

Reply:
The word size in the current VM  (EVM) is 256-bit. Implementing arbitrary precision on that VM will be tricky/costly.
Since the argument arrive as just binaries I think knowing whether it is signed or unsigned will help.

Reply:
@happi why is this the case? what do we do better then the evm?